ddtrace/context.py

```diff
@@ -26,164 +23,71 @@ class Context(object):
-        with self._lock:
-            return self._parent_span_id
-
-    @property
-    def sampling_priority(self):
-        """Return current context sampling priority."""
-        with self._lock:
-            return self._sampling_priority
-
-    @sampling_priority.setter
-    def sampling_priority(self, value):
-        """Set sampling priority."""
-        with self._lock:
-            self._sampling_priority = value
+        """Return active context span_id."""
+        return self._active_span_id
 
     def clone(self):
-        """
-        Partially clones the current context.
-        It copies everything EXCEPT the registered and finished spans.
-        """
-        with self._lock:
-            new_ctx = Context(
-                trace_id=self._parent_trace_id,
-                span_id=self._parent_span_id,
-                sampling_priority=self._sampling_priority,
-            )
-            new_ctx._current_span = self._current_span
-            return new_ctx
+        new_ctx = Context(
+            trace_id=self._trace_id,
+            span_id=self._active_span_id,
+            sampling_priority=self.sampling_priority,
+            dd_origin=self.dd_origin,
+        )
+        new_ctx._active_span_id = self._active_span_id
+        return new_ctx
 
     def get_current_root_span(self):
+        # type: () -> Optional[Span]
         """
         Return the root span of the context or None if it does not exist.
         """
-        return self._trace[0] if len(self._trace) > 0 else None
+        return self._root_span
 
     def get_current_span(self):
+        # type: () -> Optional[Span]
         """
         Return the last active span that corresponds to the last inserted
         item in the trace list. This cannot be considered as the current active
         span in asynchronous environments, because some spans can be closed
         earlier while child spans still need to finish their traced execution.
         """
-        with self._lock:
-            return self._current_span
+        return self._current_span
 
     def _set_current_span(self, span):
-        """
-        Set current span internally.
-
-        Non-safe if not used with a lock. For internal Context usage only.
-        """
         self._current_span = span
         if span:
-            self._parent_trace_id = span.trace_id
-            self._parent_span_id = span.span_id
+            self._trace_id = span.trace_id
+            self._active_span_id = span.span_id
         else:
-            self._parent_span_id = None
+            self._active_span_id = None
 
     def add_span(self, span):
-        """
-        Add a span to the context trace list, keeping it as the last active span.
-        """
-        with self._lock:
-            self._set_current_span(span)
-
-            self._trace.append(span)
-            span._context = self
+        # type: (Span) -> None
+        """Activate span in the context."""
+        if not span.parent_id:
```

Comment:
what if we are the local root, but are part of a distributed trace?

Reply:
I've tried to make the distinction much more distinct now, let me know what you think!